lib/core/test/unit/Cardano/Wallet/Api/ServerSpec.hs

```diff
@@ -265,3 +317,116 @@ errorHandlingSpec = describe "liftHandler and toServerError" $ do
 
         runHandler (getNetworkClock (error "bomb") True)
             `shouldReturn` Left expectedErr
+
+listStakeKeysSpec :: Spec
+listStakeKeysSpec = do
+    describe "Listing stake keys" $ do
+        it "no rewards => sum of all keys' stake == utxo balance" $ do
+            property prop_listStakeKeysBalance
+        it "ours and foreign keys are disjoint" $ do
+            property prop_listStakeKeysDisjoint
+
+prop_listStakeKeysDisjoint
+    :: UTxO
+    -> Set RewardAccount
+    -> (Map RewardAccount (Maybe Coin))
+    -> Property
+prop_listStakeKeysDisjoint utxo ours' m =
+    let
+        -- Build a list to pass listStakeKeys'
+        ours = zipWith (\ix acc -> (acc, ix, noDelegation))
+            [(0::Natural) ..]
+            (Set.toList ours')
+
+        ApiStakeKeys{_ours, _foreign, _none } = runIdentity $
+            listStakeKeys' utxo accountOfAddress (const $ pure m) ours
+
+        ourKeys = keys _ours
+        foreignKeys = keys _foreign
+    in
+        classify (length ourKeys > 1 && length foreignKeys > 1) "non-trivial" $
+            keys _ours `disjoint` keys _foreign
+  where
+    labeledCheck notF f a b = counterexample
+        (show a <> " " <> notF <> " " <> show b)
+        (property $ f a b)
+
+    disjoint = labeledCheck "is not disjoint from" Set.disjoint
+
+    keys = Set.fromList . map (view #_key)
+
+prop_listStakeKeysBalance
+    :: UTxO
+    -> Set RewardAccount
+    -> Property
+prop_listStakeKeysBalance utxo ours' = checkCoverage $
+    let
+        -- Build a list to pass listStakeKeys'
+        --
+        -- Passing the same account twice in the list would break the
+        -- property, but protecting against this in the implementation of
+        -- listStakeKeys' might not be worth it.
+        ours = zipWith (\ix acc -> (acc, ix, noDelegation))
+            [(0::Natural) ..]
+            (Set.toList ours')
+        ApiStakeKeys{_ours, _foreign, _none } = runIdentity $
+            listStakeKeys' utxo accountOfAddress (const $ pure mempty) ours
+
+        ourKeys = keys _ours
+        foreignKeys = keys _foreign
+        isNonTrivial =
+            length ourKeys > 1
+            && length foreignKeys > 1
+            && totalStake [_none] > 0
+
+    in
+       cover 35 isNonTrivial "non-trivial" $
+            ((totalStake _ours) +
+                (totalStake _foreign) +
+                (totalStake [_none]))
+                ===
+            getQuantity (coinToQuantity (TokenBundle.getCoin (balance utxo)))
+  where
+    keys = Set.fromList . map (view #_key)
+    totalStake = sum . map (getQuantity . view (#_stake))
```

Comment:
I'm just guessing, but perhaps it's possible to use `F.foldMap` here, if the elements are monoidal. (Though I haven't tried it.)

```suggestion
    totalStake = F.foldMap (getQuantity . view (#_stake))
```

Reply:
I recall trying it, but the problem was that they weren't `Monoidal`. Maybe if one converted back to `Coin` rather than `Quantity` though...

Reply:
I guess you could use `Semigroup.Sum` (which is an instance of `Semigroup`).

So:
```hs
    totalStake = getSum $ F.foldMap (Sum . getQuantity . view (#_stake))
```

This would have the small advantage of only traversing the list once.

But it's probably not worth it for this small traversal 😄.

Reply:
I ended up using `Coin` now.